File: server/src/engine/confirm.ts
/**
 * Confirm routine (shared) — runs on auto-accept and when a secretary approves a
 * pending request. Ported from functions/src/engine/confirm.ts; Firestore
 * batch/reads → Mongo. Idempotent: already-CONFIRMED/DONE tasks are not
 * re-awarded.
 *
 * For every filled, not-yet-confirmed task: set CONFIRMED, award its points to
 * the assignee, add it to the roster (contact-facing roles only). Then write the
 * roster onto the request, set status 'Request Accepted', send invites + emails.
 */

import { Document, WithId } from "mongodb";
import { col, oid } from "../db";
import { calendarService } from "../services/calendar";
import { emailService } from "../services/email";
import { appendActivity } from "../lib/log";
import { RosterEntry } from "../types";

const ROSTER_ROLES = new Set([
  "Event Coordinator",
  "Photographer",
  "Videographer",
  "Content Writer",
]);

export async function confirmRequest(requestId: string): Promise<void> {
  const _id = oid(requestId);
  if (!_id) return;
  const request = await col.requests().findOne({ _id });
  if (!request) return;
  const tasks = await col.tasks().find({ requestId }).toArray();

  const pointsByEmail: Record<string, number> = {};
  const roster: RosterEntry[] = [];
  const sideEffects: WithId<Document>[] = [];

  for (const t of tasks) {
    if (t.status === "UNFILLED" || !t.email) continue;

    if (ROSTER_ROLES.has(t.task)) {
      roster.push({ role: t.task, name: t.member, email: t.email, phone: t.phone || "" });
    }

    // Already confirmed/done — don't re-award or re-notify.
    if (t.status === "CONFIRMED" || t.status === "DONE") continue;

    await col.tasks().updateOne(
      { _id: t._id },
      { $set: { status: "CONFIRMED", pointsAwarded: true } }
    );
    const e = String(t.email).toLowerCase();
    pointsByEmail[e] = (pointsByEmail[e] || 0) + (t.points || 0);
    sideEffects.push(t);
  }

  for (const [email, pts] of Object.entries(pointsByEmail)) {
    await col.team().updateOne({ _id: email as never }, { $inc: { points: pts } });
  }

  await col.requests().updateOne({ _id }, { $set: { status: "Request Accepted", roster } });

  // Side effects after the state is committed.
  for (const t of sideEffects) {
    if (t.atEvent && t.eventStart && t.eventEnd) {
      await calendarService.createHold({
        email: t.email,
        title: `${t.refCode} ${t.task} — ${t.eventName}`,
        start: new Date(t.eventStart),
        end: new Date(t.eventEnd),
        description: t.venue || "",
      });
    } else if (t.deadline) {
      await calendarService.createReminder({
        email: t.email,
        title: `${t.refCode} ${t.task} due — ${t.eventName}`,
        due: new Date(t.deadline),
      });
    }
    await emailService.send({
      to: t.email,
      subject: `[Assigned] ${t.refCode} ${t.task} — ${t.eventName}`,
      text:
        `You've been assigned as ${t.task} for ${t.eventName} (${t.refCode}).\n` +
        (t.deadline ? `Deadline: ${new Date(t.deadline).toLocaleString()}\n` : ""),
    });
    await appendActivity({
      event: "confirmed",
      requestId,
      refCode: t.refCode,
      member: t.email,
      detail: `${t.task} confirmed (+${t.points || 0} pts)`,
    });
  }

  // Email the requester the assigned roster.
  await emailService.send({
    to: request.contactEmail,
    subject: `[Accepted] ${request.refCode} — ${request.eventName}`,
    text: rosterEmailText(request, roster),
  });
  await appendActivity({
    event: "accepted",
    requestId,
    refCode: request.refCode,
    actor: "engine",
    detail: `Request Accepted; ${roster.length} contact(s) in roster`,
  });
}

function rosterEmailText(request: Document, roster: RosterEntry[]): string {
  const lines = roster.map(
    (m) => `  ${m.role}: ${m.name} <${m.email}>${m.phone ? ` · ${m.phone}` : ""}`
  );
  return (
    `Your request ${request.refCode} (${request.eventName}) has been accepted.\n\n` +
    `Assigned team:\n${lines.join("\n") || "  (none)"}\n`
  );
}

File: portal/engine/confirm.py
"""
Confirming a request (docs/PRD.md §5.3).

Ported from `server/src/engine/confirm.ts`, since adapted: points are now
credited on task *completion*, not here (see engine/workflow.py's
`_award_completion_points`) — confirming only locks a task in as CONFIRMED and
adds it to the roster if it's a contact-facing role. Then the roster is written
onto the request, it moves to 'Request Accepted', and the invitations go out.

**Idempotency matters here.** Tasks already CONFIRMED or DONE are skipped, so
approving twice — or a retry after a partial failure — can never re-notify or
re-add the same roster entry twice.
"""

from __future__ import annotations

from django.db import transaction

from core.activity import log_activity
from core.constants import ROSTER_ROLES, RequestStatus, TaskStatus
from services import email as email_service

from .notify import notify_assignee


def confirm_request(request_obj) -> None:
    newly_confirmed, roster = _commit_confirmation(request_obj)

    # Side effects run only after the state change is committed, so a slow or
    # failing mail relay can't leave the database half-updated.
    for task in newly_confirmed:
        notify_assignee(task)
        log_activity(
            "confirmed",
            request_obj=request_obj,
            ref_code=task.ref_code,
            member=task.email,
            detail=f"{task.task} confirmed ({task.points or 0} pts on completion)",
        )

    email_service.send(
        request_obj.contact_email,
        f"[Accepted] {request_obj.ref_code} — {request_obj.event_name}",
        _roster_email(request_obj, roster),
    )
    log_activity(
        "accepted",
        request_obj=request_obj,
        actor="engine",
        detail=f"Request Accepted; {len(roster)} contact(s) in roster",
    )


@transaction.atomic
def _commit_confirmation(request_obj):
    """Everything that touches the database, in one transaction."""
    roster: list[dict] = []
    newly_confirmed = []

    for task in request_obj.tasks.select_for_update():
        if task.status == TaskStatus.UNFILLED or not task.email:
            continue

        if task.task in ROSTER_ROLES:
            roster.append(
                {
                    "role": task.task,
                    "name": task.member,
                    "email": task.email,
                    "phone": task.phone or "",
                }
            )

        # Already handled on an earlier run — don't re-notify.
        if task.status in (TaskStatus.CONFIRMED, TaskStatus.DONE):
            continue

        task.status = TaskStatus.CONFIRMED
        task.save(update_fields=["status"])
        newly_confirmed.append(task)

    request_obj.status = RequestStatus.ACCEPTED
    request_obj.roster = roster
    request_obj.save(update_fields=["status", "roster"])

    return newly_confirmed, roster


def _roster_email(request_obj, roster: list[dict]) -> str:
    lines = [
        f"  {entry['role']}: {entry['name']} <{entry['email']}>"
        + (f" · {entry['phone']}" if entry.get("phone") else "")
        for entry in roster
    ]
    return (
        f"Your request {request_obj.ref_code} ({request_obj.event_name}) has been accepted.\n\n"
        f"Assigned team:\n" + ("\n".join(lines) or "  (none)") + "\n"
    )
